Do not recurse on semi-local elements

permcody opened the pull request:
Instead of recursing if elements (or nodes) that are semilocal, this patch
allows recursion on neighbors of active local elements but nothing
beyond that. I have verified that this at least removes the previous
assertion (seg fault) on the attached 3D input file. This is way
to large to turn into a regression test. More work to manufacture a
good test is needed.

closes #7200

Tagging @jwpeterson since this is his idea
@roystgnr and @dschwen.

permcody committed:
Do not recurse on semi-local elements

Instead of recursing if elements (or nodes) are semilocal this patch
allows recursion on neighbors of active local elements but nothing
beyond that. I have verified that this at least removes the previous
assertion (seg fault) on the attached 3D input file. This is way
to large to turn into a regression test. More work to manufacture a
good test is needed.

closes #7200

jwpeterson commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -636,7 +610,7 @@ FeatureFloodCount::updateFieldInfo()
       // Loop over the ghosted ids to update cells with ghost information
       for (auto entity : feature._ghosted_ids)
-        _ghosted_entity_ids[entity] = feature_number;
+        _ghosted_entity_ids[entity] = 1;
```
Are you sure it's not "4"?

permcody replied:
Yeah, I'm sure...

I changed this line because I really don't want to see a "multi-colored" field when I look at ghosted ids. Using processor_id might be interesting but certainly not something related to the feature or grain. Just using a constant gives the "expected" looking field.

permcody commented:
The halos don't look quite right.... I think the problem is that I need to always premark entries (like I was doing before) even when I don't do the recursion step. I'll experiment with that some more before we merge this.

permcody committed:
New input files

permcody committed:
Completing all of the edge cases with the new flood algorithm

permcody committed:
Improved mergeSets algorithm

permcody commented:
Got a little carried away and spent about 8 hours rewriting everything again (on Saturday). I was able to create a few pretty good regression problems that run in several second and do advanced remapping so I'll even be able to create better regression tests from all of this. I need to polish up all the inputs and also I need to try this on the 3D adaptive problem before this is merged. I'll do all that on Monday.

permcody commented on framework/include/restart/DataIO.h:
```diff
@@ -374,7 +374,7 @@ dataLoad(std::istream & stream, std::list<T> & l, void * context)
     T data;
     loadHelper(stream, data, context);
-    l.push_back(data);
+    l.push_back(std::move(data));
```
I need to go through all the DataIO methods and mark many of them with `std::move`...

permcody replied:
Well I went through DataIO and we only needed one other `std::move()`

permcody committed:
Use halo entities instead of inflation

Instead of inflating the bounding boxes around the
marked entities. I'm switching to using the level 1
halos.

permcody committed:
A little cleanup and a better test problem

permcody committed:
Several C++11 enhancements

permcody committed:
Avoid copies in some of the DataIO routines

permcody commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -727,56 +679,57 @@ FeatureFloodCount::visitElementalNeighbors(const Elem * elem, unsigned long curr
+          feature->_halo_ids.insert(neighbor->id());
+
+          flood(neighbor, current_idx, feature);
+        }
```
@roystgnr - Take a look at this condition for recursing. It's actually quite simple now and much more elegant that looking at the semi-local set. Also seems to work just fine :)

permcody committed:
New Remapping test
Fixing up nodal test

permcody committed:
Making FeatureData more object oriented

permcody commented:
This requires a one line change to BISON. I have that patch ready.

permcody committed:
Make the remapping handle very high pressure cases

permcody commented:
Sorry this ended up being huge @jwpeterson - It should be ready to review and merge. I have the BISON patch ready.

jwpeterson commented on framework/src/restart/DataIO.C:
```diff
@@ -218,13 +218,11 @@ dataLoad(std::istream & stream, std::string & v, void * /*context*/)
-  v = s;
-  delete[] s;
+  // Read the string
+  stream.read(&v[0], sizeof(char) * size);
```
Just found out today that this is now *guaranteed* to work in C++11.

roystgnr replied:
Wait, what changed in C++11?  I thought &v[0] is guaranteed to work iff the vector (in this case string?) is non-empty, regardless of standard.  C++11 added http://en.cppreference.com/w/cpp/container/vector/data to make handling the possibly-empty cases easier.

jwpeterson replied:
I was referring to [this](http://stackoverflow.com/questions/16425338/is-it-safe-to-pass-stdstring-to-c-style-apis) which says that C++03 did not guarantee strings were contiguous in memory (although, how could they *not* be?) but can't vouch for its accuracy or my reading of it.

jwpeterson commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -65,14 +66,14 @@ class FeatureFloodCount :
 
   /// This enumeration is used to indicate status of the grains in the _unique_grains data structure
-  enum STATUS
+  enum class Status
```
I hope all our minimum required compilers support `enum class`, this is another thing we never tested.

jwpeterson commented on modules/phase_field/include/postprocessors/FeatureFloodCount.h:
```diff
@@ -220,20 +281,20 @@ class FeatureFloodCount :
    */
   template<class InputIterator>
-  inline bool setsIntersect(InputIterator first1, InputIterator last1, InputIterator first2, InputIterator last2) const
+  static inline bool setsIntersect(InputIterator first1, InputIterator last1, InputIterator first2, InputIterator last2)
```
Future note: we should eventually move this to MooseUtils.

jwpeterson commented on modules/phase_field/src/postprocessors/FauxGrainTracker.C:
```diff
@@ -34,16 +34,23 @@ FauxGrainTracker::~FauxGrainTracker()
-FauxGrainTracker::getEntityValue(dof_id_type entity_id, FeatureFloodCount::FIELD_TYPE field_type, unsigned int var_idx) const
+FauxGrainTracker::getEntityValue(dof_id_type entity_id, FeatureFloodCount::FieldType field_type, unsigned int var_idx) const
 {
+  auto use_default = false;
```
LOL, `auto` vs. `bool`.  OK bro

jwpeterson commented on modules/phase_field/src/postprocessors/FeatureFloodCount.C:
```diff
@@ -467,129 +468,76 @@ FeatureFloodCount::mergeSets(bool use_periodic_boundary_info)
+   * items into a vector of vectors instead.
+   */
+  _feature_count = 0;
+  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
```
```
auto map_num = decltype(_maps_size)(0)
```
vs.
```
unsigned map_num = 0;
```
seems ridiculous to me.

friedmud replied:
It's not ridiculous.  The type of `map_num` will now always stay in sync with the type of `_maps_size`... so you don't get into situations where `_maps_size` is changed to a `long unsigned int` but you forget to change `map_num` and your code is broken for large problems...

jwpeterson replied:
The same problem is solved with a typedef.

friedmud replied:
That's way worse.  You're promoting us to have millions of typedefs running around for each little thing?

I don't want to have to remember to use `some_other_index_type` for each line of code...

This code is explicit and self-explanatory.

jwpeterson replied:
> I don't want to have to remember to use some_other_index_type for each line of code...

You're already doing that, it's just spelled "auto foo = decltype(bar)(0)" instead.

friedmud replied:
Nope - that line of code is the same regardless of what's on the right hand side.

vs having to remember "Oh yeah... this thing is a `dof_id_type`, this thing is a `element_id_type`, this thing is a `blah_id_type`"

friedmud replied:
And the problem is... if you screw up and assign a `bar_id_type` to a `foo_id_type` they might be the same right now and not give any compiler warnings and work just fine so you would never know... and then later you change one of them and you have a bug.

_This_ line of code is foolproof.

jwpeterson replied:
OK, I think I can see your point, but the syntax is so ugly... I just can't get behind it.

friedmud replied:
Yeah - I'm not the biggest fan of the syntax either.  We (me @permcody and @dschwen   ) already talked about this construction before.  I personally prefer:

`decltype(foo) bar = 0`

That is valid syntax and matches up better with what we're used to.  The problem is that sometimes `foo` is `const`... so this would actually make `bar` const as well and then you get a compile error on the `bar++`.

`auto bar = decltype(foo)(0)`

Doesn't have that issue.

You could do:

`decltype(std::remove_const(foo)) bar = 0` or somesuch... but that's deeply unsatisfying as well ;-)

`auto bar = decltype(foo)(0)` gets the job done and is simple (enough).

permcody replied:
One of these days I'll figure out how to get rid of this whole "maps"
business so I can just use a range based for. Then I won't have this
problem. There are always going to be cases though were people need to
iterate over a range of integers and C++11 doesn't provide any clean ways
to just do that with an auto type...

On Wed, Jun 22, 2016 at 2:45 PM Derek Gaston <notifications@github.com>
wrote:

> In modules/phase_field/src/postprocessors/FeatureFloodCount.C
> <https://github.com/idaholab/moose/pull/7201#discussion_r68131634>:
>
> >
> > -      _partial_feature_sets[rank][map_num].clear();
> > +  /**
> > +   * All of the merges are complete and stored in a vector of lists. To make several
> > +   * of the sorting and tracking algorithms more straightforward, we will move these
> > +   * items into a vector of vectors instead.
> > +   */
> > +  _feature_count = 0;
> > +  for (auto map_num = decltype(_maps_size)(0); map_num < _maps_size; ++map_num)
>
> Yeah - I'm not the biggest fan of the syntax either. We (me @permcody
> <https://github.com/permcody> and @dschwen <https://github.com/dschwen> )
> already talked about this construction before. I personally prefer:
>
> decltype(foo) bar = 0
>
> That is valid syntax and matches up better with what we're used to. The
> problem is that sometimes foo is const... so this would actually make bar
> const as well and then you get a compile error on the bar++.
>
> auto bar = decltype(foo)(0)
>
> Doesn't have that issue.
>
> You could do:
>
> decltype(std::remove_const(foo)) bar = 0 or somesuch... but that's deeply
> unsatisfying as well ;-)
>
> auto bar = decltype(foo)(0) gets the job done and is simple (enough).
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/idaholab/moose/pull/7201/files/1800330b4de20f90b511d453a72308208f05e5bb#r68131634>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AC5XIJjey9pMcPUHk0pcCshUiD9OixJ2ks5qOZ7YgaJpZM4I34-f>
> .
>

friedmud replied:
Here's an interesting related read: https://herbsutter.com/2013/08/12/gotw-94-solution-aaa-style-almost-always-auto/

It's kind of where I'm headed...

roystgnr replied:
Wait - Herb Sutter is still writing Guru of the Week?  I must have a decade of posts to catch up on...

jwpeterson commented:
:+1:
